Refactor/reaction-realtime-update

Changyu-Ryou opened the pull request:
## Linked Issue
close #153 

## 공유할 사항
- 리액션 추가/제거
- 실시간 다른 유저의 리액션 상태 업데이트

## 논의할 사항
- 없습니다. ❌

Changyu-Ryou committed:
Refactor: update reaction update service

db reaction 상태를 update하는 service의 로직을 refactoring

Changyu-Ryou committed:
Feat: add Backend socket reaction update

백엔드에서 소켓으로 전달받은 데이터를 이용해 리액션 상태 업데이트 기능 추가

Changyu-Ryou committed:
Refactor: remove dead code

불필요한 코드 제거

Changyu-Ryou committed:
Feat: add updateReaction util

프론트에서 소켓에 리액션 업데이트 정보를 담아 전달하는 유틸 구현

Changyu-Ryou committed:
Refactor: update props data

props로 전달해주는 데이터에 chatId 추가

Changyu-Ryou committed:
Feat: add Frontend socket update reaction

프론트에서 소켓을 통해 리액션 업데이트 정보를 받아 message의 state를 변경하는 기능 구현

Changyu-Ryou committed:
Feat: add Reaction update event on ActionBar

ActionBar에서 리액션을 업데이트하는 이벤트 추가

Changyu-Ryou committed:
Feat: add Reaction update event on AddReactionButton

AddReactionButton을 통해 리액션 상태를 업데이트하는 이벤트 핸들러 추가

Changyu-Ryou committed:
Refactor: update ThreadReactionCard component

ThreadReactionCard에서 상태 변화에 따라 색상을 반영하고 ThreadReactionCard 클릭을 통해 리액션 업데이트 이벤트를 발생시킬 수 있도록 변경

Changyu-Ryou committed:
Refactor: update hover ActionBar event

chatMessage에 hover시 혹은 이모지 모달이 열렸을 경우에만 ActionBar 컴포넌트가 렌더링 되도록 변경

Changyu-Ryou committed:
Refactor: remove dead code

불필요한 코드 제거

Changyu-Ryou committed:
Fix: ThreadReactionCard bug fix

ThreadReactionCard에서 _id를 찾지 못해 발생하는 버그 수정

rockpell commented on frontend/src/organism/ActionBar/ActionBar.js:
```diff
@@ -12,14 +13,24 @@ import {
+    updateReaction({
+      workspaceUserInfo,
+      socket,
+      emoji: emoji.native || emoji,
```
emoji.native와 emoji는 어떤 차이가 있는건가요?

Changyu-Ryou replied:
emoji-mart 모달에서 이벤트가 발생했을 경우에는 emoji.native에서 string을 받아 오게 되구요, emoji-mart 모달이 아닌 일반 이모티콘 버튼을 눌러 이벤트를 작동시켰다면 따로 emoji.native가 없기때문에 그냥 전달받은 emoji를 다시 전달하도록 했습니다.

rockpell commented on frontend/src/organism/ChatRoom/ChatRoom.js:
```diff
@@ -60,16 +60,61 @@ const ChatRoom = () => {
+          message.reactions.map((item, idx) => {
+            if (item.emoji === reaction.emoji) {
+              if (reaction.type) {
+                const userInfo = [
+                  {
+                    _id: reaction.workspaceUserInfoId,
+                    displayName: reaction.displayName,
+                  },
+                ]
+                item.users = [...item.users, ...userInfo]
```
array로 선언 후 item.users에 ... 연산자로 꺼내서 넣어주는 작업에 대해서 이해가 잘 가지 않습니다.
userInfo에 하나의 object만 들어가는것 같은데 바로 object로 넣는건 불가능할까요?

Changyu-Ryou replied:
그러게요 이 부분은 삽질을 좀 하다가 이렇게 된 것 같은데 수정이 필요하겠네요

rockpell reviewed (approved):
수고하셨습니다!

bell-won commented on backend/service/reaction.js:
```diff
@@ -1,20 +1,21 @@
 import { verifyRequiredParams, dbErrorHandler } from '../util'
 
-const addReaction = async ({ chatId, workspaceUserInfoId, emoticon }) => {
+const updateReaction = async ({ chatId, workspaceUserInfoId, emoticon }) => {
```
add reaction, remove reaction으로 작성하는게 더 restfull한 방법이 아닐까 싶습니당

Changyu-Ryou replied:
이 부분도 분리하는걸로 해서 다시 변경해보겠습니다.

bell-won commented on frontend/src/util/updateReaction.js:
```diff
@@ -0,0 +1,20 @@
+function updateReaction({
```
util 함수로 분리한 이유가 있을까요??

Changyu-Ryou replied:
정확히는 util이라기 보다 여러 컴포넌트에 사용되어져 중복되는 코드를 줄이려고 분리해두었습니다.

bell-won commented on frontend/src/organism/ChatRoom/ChatRoom.js:
```diff
@@ -60,16 +60,61 @@ const ChatRoom = () => {
     }
     return () => {
-      socket && socket.off('new message')
+      socket && socket.off('new message') && socket.off('update reaction')
```
socket.off 함수의 결과값이 있나요?? 이렇게도 쓸 수 있는지 몰랐네용

Changyu-Ryou replied:
요 부분은 확인이 필요할 것 같네요

bell-won commented on frontend/src/organism/ChatRoom/ChatRoom.js:
```diff
@@ -60,16 +60,61 @@ const ChatRoom = () => {
 
+  const chageReactionState = (messages, reaction) => {
+    let done = false
+    const arr = messages.map((message, idx) => {
```
arr이 재사용되는게 아니라면
return message.map으로 해도 되지않을까요??

Changyu-Ryou replied:
넵 그렇게 해도 문제 없을 것 같습니다

bell-won reviewed (commented):
고생하셨습니당

Changyu-Ryou committed:
Refactor: update backend removeReaction and addReaction service

기존 updateReaction 한개로 리액션을 업데이트 했던 부분을 removeReaction과 addReaction으로 분리해 restful 하게 변경했습니다.

Changyu-Ryou committed:
Refactor: remove updateReaction util

분리한 updateReaction util을 제거

Changyu-Ryou committed:
Refactor: update reviewed ChatRoom code

리뷰받은 코드 수정
- return value를 Array.map 결과 값으로 설정
- socket.off 코드 수정
- reaction update 예외처리 추가

Changyu-Ryou committed:
Refactor: seperate updateReaction logic

updateReaction을 리액션 추가인지 삭제인지에 따라 type을 지정해 socket 통신 하도록 변경
